File: cpp/somme.cpp
/* gator 2017
input: 1 5
output: 1
*/

#include <iostream>
#include <fstream>

int sequenza(int n, bool pollatz, int tCollatz = -1)
{
	int m = ((pollatz) ? 5 : 3), tot = 1;
	
	auto pari = [] (int n) {
		return ((n % 2) == 0) ? true : false;
	};
		

	while(n != 1) {		
		n = (pari(n)) ? n/2 : n*m+1;
		
		tot++;
		
		if(tCollatz != -1 && tot > tCollatz)
			break;
	}
	
	return tot;
}

int main()
{
	std::ifstream in("input.txt");
	std::ofstream out("output.txt");
	
	int N1, N2, tot = 0, collatz;
	in >> N1 >> N2;
	
	for(int i = N1; i <= N2; i++) {
		collatz = sequenza(i, false);
		if(sequenza(i, true, collatz) < collatz)
			tot++;
	}
	
	out << tot;
	
	out.close();
	in.close();
	
	return 0;
}

File: python/somme.py
def sequenza(n, pollatz, tCollatz = -1):
	tot = 1
	
	if pollatz:
		m = 5
	else:
		m = 3
		
	pari = lambda n: ((n % 2) == 0)

	while n != 1:
		if pari(n):
			n /= 2
		else:
			n = n*m+1
			
		tot += 1
		
		if tCollatz != -1 and tot > tCollatz:
			break
			
	
	return tot


with open('input.txt') as fin:
	N = fin.readline().split(' ')

	
N[0] = int(N[0])
N[1] = int(N[1])

tot = 0

for i in range(N[0], N[1]+1):
	collatz = sequenza(i, False)
	
	if sequenza(i, True, collatz) < collatz:
		tot += 1
		
with open('output.txt', 'w') as fout:
	fout.write(str(tot))